core/arch/riscv64/emit_utils.c

```diff
@@ -587,14 +587,255 @@ emit_inline_ibl_stub(dcontext_t *dcontext, byte *pc, ibl_code_t *ibl_code,
+     *     TLS_REG3_SLOT: scratch space
+     * There are following entries with the same context:
+     *     indirect_branch_lookup
+     *     unlink_stub_entry
+     * target_delete_entry:
+     *     a0: scratch
+     *     a1: table entry pointer from ibl lookup hit path
+     *     a2: app's a2
+     *     TLS_REG0_SLOT: app's a0
+     *     TLS_REG1_SLOT: app's a1
+     *     TLS_REG2_SLOT: app's a2
+     * On miss exit we output:
+     *     a0: the dcontext->last_exit
+     *     a1: jr a1
+     *     a2: app's a2
+     *     TLS_REG0_SLOT: app's a0 (recovered by fcache_return)
+     *     TLS_REG1_SLOT: app's a1 (recovered by fcache_return)
+     * On hit exit we output:
+     *     a0: fragment_start_pc (points to the fragment prefix)
+     *     a1: scratch reg
+     *     a2: app's a2
+     *     TLS_REG0_SLOT: app's a0 (recovered by fragment_prefix)
+     *     TLS_REG1_SLOT: app's a1 (recovered by fragment_prefix)
+     */
+
+    /* Spill a0. */
+    APP(&ilist, instr_create_save_to_tls(dc, DR_REG_A0, TLS_REG3_SLOT));
+
+    /* Load the hash mask into scratch register a1, which will be used in the hash
+     * function.
+     */
+    APP(&ilist,
+        INSTR_CREATE_ld(dc, opnd_create_reg(DR_REG_A1),
+                        opnd_create_base_disp(dr_reg_stolen, DR_REG_NULL, 0,
+                                              TLS_MASK_SLOT(ibl_code->branch_type),
+                                              OPSZ_8)));
+
+    /* Memory barrier for the hash mask. We need a barrier to ensure we see updates
+     * properly.
+     * fence rw, rw
+     */
+    APP(&ilist,
+        INSTR_CREATE_fence(dc, opnd_create_immed_int(0x3, OPSZ_4b),
+                           opnd_create_immed_int(0x3, OPSZ_4b),
+                           opnd_create_immed_int(0x0, OPSZ_4b)));
+
+    /* Load lookup table base into a0. */
+    APP(&ilist,
+        INSTR_CREATE_ld(dc, opnd_create_reg(DR_REG_A0),
+                        opnd_create_base_disp(dr_reg_stolen, DR_REG_NULL, 0,
+                                              TLS_TABLE_SLOT(ibl_code->branch_type),
+                                              OPSZ_8)));
+
+    /* The hash "function":
+     * a1: hash mask
+     * a2: indirect branch target
+     */
+    APP(&ilist,
+        INSTR_CREATE_and(dc, opnd_create_reg(DR_REG_A1), opnd_create_reg(DR_REG_A1),
+                         opnd_create_reg(DR_REG_A2)));
+
+    /* Now, a1 holds the hash table index, use slli+add to get the table entry. */
+    ASSERT(4 - HASHTABLE_IBL_OFFSET(ibl_code->branch_type) >= 0);
+    if (4 - HASHTABLE_IBL_OFFSET(ibl_code->branch_type) > 0) {
+        APP(&ilist,
+            INSTR_CREATE_slli(
+                dc, opnd_create_reg(DR_REG_A1), opnd_create_reg(DR_REG_A1),
+                opnd_add_flags(
+                    opnd_create_immed_int(4 - HASHTABLE_IBL_OFFSET(ibl_code->branch_type),
+                                          OPSZ_6b),
+                    DR_OPND_IMM_PRINT_DECIMAL)));
```

Comment:
Yes, there is still a lot that can be improved on the decoder and encoder, including this one and the one you pointed out below. My plan is to push the already stable code upstream first, and then improve these details. Currently, the upstream code still cannot run any program, which also results in most tests being unable to enable RISC-V.

Reply:
OK, are these issues written down somewhere, maybe in the #3544 container case?  Just want to not lose the action item.

Reply:
Could we also have a TODO i#3544 comment here so the reader knows that we realize this is not ideal and not what we want long-term.  Otherwise new developers will see this and think this is the golden standard to use and not know this is not supposed to be how it looks long-term.